src/os/file_anyos.go

```diff
@@ -97,6 +97,11 @@ type unixFileHandle uintptr
 // read and any error encountered. At end of file, Read returns 0, io.EOF.
 func (f unixFileHandle) Read(b []byte) (n int, err error) {
 	n, err = syscall.Read(syscallFd(f), b)
+	// In case of EISDIR, n == -21.
+	// This breaks the assumption that n always represent the number of read bytes.
+	if err == syscall.EISDIR {
+		n = 0
+	}
```

Comment:
I am confused by this change. I tried [1] to recreate your described behavior on my linux machine but could not observe, that the read syscall would return `-EISDIR` for the bytes read when this error is encountered.

[1]: https://gist.github.com/leongross/790669a0e83369970f9e7b9bdff30bda

Reply:
I suspect `n` should be set to 0 for any error.
This might actually be a bug in the Syscall function (implemented in the compiler).

Reply:
I am using llvm-18 on debian 12 (kernel 6.1)
Let me see if I can reproduce in C, I only encountered this when compiling go code so far.

Reply:
I have updated the commit to mention -1 instead of -21.
Fix stays the same, in case of `EISDIR` we can force `n = 0`

Reply:
I think the proper thing to do here is to check the `err` returned in https://github.com/noboruma/tinygo/blob/fix-os-unix-read/src/os/file.go#L471 and then abort the operation. If there really is an error, we should abort anyway, so the value of `n` should not be read anyway.

Reply:
@leongross we can, this was actually the first fix I did. 
However this code is a copy-paste from official Golang source (GC), so any modification here means we are forking away from std. In std library they handle the `n` issue at a lower layer, as proposed here.